lib/Dialect/FIRRTL/Transforms/IMDeadCodeElim.cpp

```diff
@@ -223,11 +250,19 @@ void IMDeadCodeElimPass::runOnOperation() {
   auto circuit = getOperation();
   instanceGraph = &getAnalysis<InstanceGraph>();
 
+  // Create a vector of modules in the post order of instance graph.
+  // FIXME: We copy the list of modules into a vector first to avoid iterator
+  // invalidation while we mutate the instance graph. See issue 3387.
+  SmallVector<FModuleOp, 0> modules(llvm::make_filter_range(
+      llvm::map_range(
+          llvm::post_order(instanceGraph),
+          [](auto *node) { return dyn_cast<FModuleOp>(*node->getModule()); }),
+      [](auto module) { return module; }));
+
   // Forward constant output ports to caller sides so that we can eliminate
   // constant outputs.
-  for (auto *node : llvm::post_order(instanceGraph))
-    if (auto module = dyn_cast_or_null<FModuleOp>(*node->getModule()))
-      forwardConstantOutputPort(module);
+  for (auto module : modules)
```

Comment:
Is this not adequately done in IMCP?

Reply:
Right. It was added to propagate constants outputs introduce by canonicalizer after IMCP run. IIRC it was required for sifive 1.5 release.